### PR TITLE
JAMES-3050 Distributed James procedures 04: ElasticSearch indexing

### PR DESCRIPTION
**Pull request by Arsnael**

### CODE REVIEW

File: src/site/markdown/server/manage-guice-distributed-james.md [hunk 1]
```diff
@@ -204,3 +205,46 @@ reprocessing all the failed events registered in event dead letters.
...
+ - Now trigger a [reindex](https://www.elastic.co/guide/en/elasticsearch/reference/6.3/docs-reindex.html)
+from the old index to the new one
+ - When this is done, add the `mailboxReadAlias` alias to the new index
+ - Now that the migration to the new index is done, you can 
+[drop the old index](https://www.elastic.co/guide/en/elasticsearch/reference/6.3/indices-delete-index.html)
+ - You might want as well modify the James configuration file 
+[elasticsearch.properties](https://github.com/apache/james-project/blob/master/dockerfiles/run/guice/cassandra-rabbitmq/destination/conf/elasticsearch.properties)
+by setting the parameter `elasticsearch.index.mailbox.name` to the name of your new index. This is to avoid that James re-creates
+the old index as soon as you restart it for any reason. You can find more information about it by reading
+[ElasticSearch configuration](config-elasticsearch.html)
```
Comment (chibenwa) [inline src/site/markdown/server/manage-guice-distributed-james.md]: A would have had putted this `You can find more information about it by reading
[ElasticSearch configuration](config-elasticsearch.html)` in the first few lines of the ElasticSearch indexing section

File: src/site/markdown/server/manage-guice-distributed-james.md [hunk 2]
```diff
@@ -204,3 +205,46 @@ reprocessing all the failed events registered in event dead letters.
...
+so that now James writes on the old and new indexes, while only keeping reading on the first one
+ - Now trigger a [reindex](https://www.elastic.co/guide/en/elasticsearch/reference/6.3/docs-reindex.html)
+from the old index to the new one
+ - When this is done, add the `mailboxReadAlias` alias to the new index
+ - Now that the migration to the new index is done, you can 
+[drop the old index](https://www.elastic.co/guide/en/elasticsearch/reference/6.3/indices-delete-index.html)
+ - You might want as well modify the James configuration file 
+[elasticsearch.properties](https://github.com/apache/james-project/blob/master/dockerfiles/run/guice/cassandra-rabbitmq/destination/conf/elasticsearch.properties)
+by setting the parameter `elasticsearch.index.mailbox.name` to the name of your new index. This is to avoid that James re-creates
+the old index as soon as you restart it for any reason. You can find more information about it by reading
```
Comment (chibenwa) [inline src/site/markdown/server/manage-guice-distributed-james.md]: to avoid that James recreates index upon restart

Comment (chibenwa) [reply src/site/markdown/server/manage-guice-distributed-james.md]: Shouldn't we have a way to disable index creation upon startup?

File: src/site/markdown/server/manage-guice-distributed-james.md [hunk 3]
```diff
@@ -204,3 +205,46 @@ reprocessing all the failed events registered in event dead letters.
...
+[reIndex all mails](manage-webadmin.html#ReIndexing_all_mails),
+[reIndex mails from a mailbox](manage-webadmin.html#ReIndexing_a_mailbox_mails)
+or even just [reIndex a single mail](manage-webadmin.html#ReIndexing_a_single_mail).
+
+When checking the result of a reIndexing task, you might have failed reprocessed mails. You can still use the task ID to
+[reprocess previously failed reIndexing mails](manage-webadmin.html#Fixing_previously_failed_ReIndexing).
+
+But despite all those tools, sometimes you might still need to depend on the API provided by ElasticSearch itself. 
+Let's say for example that you want to scale up your nodes, or increase your shard count or replication factor on your index, 
+or even update your language analyser from English to French. In order to achieve such a procedure, you need to:
```
Comment (chibenwa) [inline src/site/markdown/server/manage-guice-distributed-james.md]: I would maintain the list of use cases separately:

```
Cases when an administrator might want to update index settings includes:
 - Scaling out: increasing the shard count might be needed.
 - Changing string analysers, for instance to target another language
 - etc..
```

File: src/site/markdown/server/manage-guice-distributed-james.md [hunk 4]
```diff
@@ -204,3 +205,46 @@ reprocessing all the failed events registered in event dead letters.
...
+
+ - [Create the new index](https://www.elastic.co/guide/en/elasticsearch/reference/6.3/indices-create-index.html) with the right
+settings and mapping
+ - James uses two aliases on the mailbox index: one for reading (`mailboxReadAlias`) and one for writing (`mailboxWriteAlias`).
+First [add an alias](https://www.elastic.co/guide/en/elasticsearch/reference/6.3/indices-aliases.html) `mailboxWriteAlias` to that new index,
+so that now James writes on the old and new indexes, while only keeping reading on the first one
+ - Now trigger a [reindex](https://www.elastic.co/guide/en/elasticsearch/reference/6.3/docs-reindex.html)
```
Comment (chibenwa) [inline src/site/markdown/server/manage-guice-distributed-james.md]: We should state that this actively relies on `_source` field being present ;-)

File: src/site/markdown/server/manage-guice-distributed-james.md [hunk 5]
```diff
@@ -204,3 +205,46 @@ reprocessing all the failed events registered in event dead letters.
...
+from the old index to the new one
+ - When this is done, add the `mailboxReadAlias` alias to the new index
```
[the remark this answers is not in the document]
Comment (Arsnael) [reply src/site/markdown/server/manage-guice-distributed-james.md]: Logically after when you delete the old index, it deletes the aliases with it? Or I shouldn't assume those?

File: src/site/markdown/server/manage-guice-distributed-james.md (hunk 2 above)
Comment (Arsnael) [reply src/site/markdown/server/manage-guice-distributed-james.md]: Yes... need to create a ticket about it, but shouldn't mention it in the doc as long as it's not implemented

File: src/site/markdown/server/manage-guice-distributed-james.md [hunk 6]
```diff
@@ -204,3 +205,46 @@ reprocessing all the failed events registered in event dead letters.
...
+under ElasticSearch. As explained in the [Mailbox Event Bus](#mailbox-event-bus) section, processing those events can
+fail sometimes.
+
+Currently, an administrator can monitor indexation failures through `ERROR` log review. You can as well
+[list failed events](manage-webadmin.html#Listing_failed_events) by looking with the group called 
+`org.apache.james.mailbox.elasticsearch.events.ElasticSearchListeningMessageSearchIndex$ElasticSearchListeningMessageSearchIndexGroup`.
+A first on-the-fly solution could be to just 
+[redeliver those group events with event dead letter](manage-webadmin.html#Redeliver_group_events).
+
+But maybe for some reasons you just lost data in ElasticSearch. Then you might need to use our WebAdmin reIndexing tasks.
```
[the remark this answers is not in the document]
Comment (Arsnael) [reply src/site/markdown/server/manage-guice-distributed-james.md]: Your data got corrupted, the disk was full, disk failure, server migration with incomplete data... better ideas?

File: src/site/markdown/server/manage-guice-distributed-james.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (chibenwa) [reply src/site/markdown/server/manage-guice-distributed-james.md]: I don't know

File: src/site/markdown/server/manage-guice-distributed-james.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (tnstung) [reply src/site/markdown/server/manage-guice-distributed-james.md]: Human mistakes, storage error but the latest snapshot is corrupted or took few days/ hours ago...

[the remark this answers is not in the document]
Comment (chibenwa) [reply src/site/markdown/server/manage-guice-distributed-james.md]: I would state `If the event storage in dead-letters fails (for instance in the face of Cassandra exceptions), then you might need to use our WebAdmin reIndexing tasks.`

File: src/site/markdown/server/manage-guice-distributed-james.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (Arsnael) [reply src/site/markdown/server/manage-guice-distributed-james.md]: Well it did when we did it with the ops... But if you prefer to be sure I can add it

[the remark this answers is not in the document]
Comment (chibenwa) [reply src/site/markdown/server/manage-guice-distributed-james.md]: Then let's keep it as is.

File: src/site/markdown/server/manage-guice-distributed-james.md [hunk 7]
```diff
@@ -205,3 +206,55 @@ reprocessing all the failed events registered in event dead letters.
...
+When checking the result of a reIndexing task, you might have failed reprocessed mails. You can still use the task ID to
+[reprocess previously failed reIndexing mails](manage-webadmin.html#Fixing_previously_failed_ReIndexing).
+
+### On the fly ElasticSearch Index setting update
+
+Sometimes you might need to update index settings. Cases when an administrator might want to update index settings include:
+
+ - Scaling out: increasing the shard count might be needed.
+ - Changing string analysers, for instance to target another language
+ - etc..
```
Comment (rouazana) [inline src/site/markdown/server/manage-guice-distributed-james.md]: s/etc../etc.

File: src/site/markdown/server/manage-guice-distributed-james.md [hunk 8]
```diff
@@ -205,3 +206,55 @@ reprocessing all the failed events registered in event dead letters.
...
+ - Now trigger a [reindex](https://www.elastic.co/guide/en/elasticsearch/reference/6.3/docs-reindex.html)
+from the old index to the new one (this actively relies on `_source` field being present)
+ - When this is done, add the `mailboxReadAlias` alias to the new index
+ - Now that the migration to the new index is done, you can 
+[drop the old index](https://www.elastic.co/guide/en/elasticsearch/reference/6.3/indices-delete-index.html)
+ - You might want as well modify the James configuration file 
+[elasticsearch.properties](https://github.com/apache/james-project/blob/master/dockerfiles/run/guice/cassandra-rabbitmq/destination/conf/elasticsearch.properties)
+by setting the parameter `elasticsearch.index.mailbox.name` to the name of your new index. This is to avoid that James re-creates
+index upon restart
+
```
Comment (rouazana) [inline src/site/markdown/server/manage-guice-distributed-james.md]: missing there is a missing work somewhere on the fact that reindexing can be very long

File: src/site/markdown/server/manage-guice-distributed-james.md [hunk 9]
```diff
@@ -205,3 +206,56 @@ reprocessing all the failed events registered in event dead letters.
 If for some other reason you don't need to redeliver all events, you have more fine-grained operations allowing you to
 [redeliver group events](manage-webadmin.html#Redeliver_group_events) or even just
 [redeliver a single event](manage-webadmin.html#Redeliver_a_single_event).
+
+## ElasticSearch Indexing
+
+A projection of messages is maintained in ElasticSearch via a listener plugged into the mailbox event bus in order to enable search features. 
+As explained in the [Mailbox Event Bus](#mailbox-event-bus) section, processing those events can fail sometimes.
```
Comment (chibenwa) [inline src/site/markdown/server/manage-guice-distributed-james.md]: IMO this should be moved in `Usual troubleshooting procedures`

**Comment by Arsnael** [comment]
test this please

**Comment by Arsnael** [comment]
test this please

**Comment by Arsnael** [comment]
test this please

**Comment by chibenwa** [comment]
No need to triger build (limited resource) on documentation contributions.

**Comment by Arsnael** [comment]
Merged

**Pull request closed without merging**